File: changelog_builder/version_control/svn.py
from collections import OrderedDict
import os
import subprocess
from changelog_builder.version_control.abstract_vcs import AbstractVcs


class Svn(object):

    def get_log(self, path, branch, min_revision):
        pr = subprocess.Popen("svn log -g -r " + str(min_revision) + ":HEAD " + path + branch, cwd=os.path.dirname(os.path.realpath(__file__)),
                              shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        (out, error) = pr.communicate()

        if error:
            raise Exception(error)

        return out

    def get_min_revision(self, path, old_branch, new_branch):
        old_revisions = self.get_revisions_for_branch(path, old_branch)
        new_revisions = self.get_revisions_for_branch(path, new_branch)

        revision_diff = list(set(new_revisions) - set(old_revisions))

        return min(revision_diff) - 1

    def get_revisions_for_branch(self, path, branch):
        pr = subprocess.Popen("svn propget svn:mergeinfo " + path + branch, cwd=os.path.dirname(os.path.realpath(__file__)),
                              shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        (out, error) = pr.communicate()

        if error:
            raise Exception(error)

        revisions = []
        branch_infos = out.split("\n")
        for branch_info in branch_infos:
            if len(branch_info) == 0:
                continue;

            branch_data = branch_info.split(":")
            revision_data = branch_data[1].split(",")
            for revision_string in revision_data:
                if not "-" in revision_string:
                    revisions.append(int(revision_string))
                else:
                    lower_end_revision = revision_string.split("-")
                    for x in range(int(lower_end_revision[0]), int(lower_end_revision[1]) + 1):
                        revisions.append(x)

        return list(OrderedDict.fromkeys(revisions))

AbstractVcs.register(Svn)

if __name__ == '__main__':
    print 'Subclass:', issubclass(Svn, AbstractVcs)
    print 'Instance:', isinstance(Svn(), AbstractVcs)

File: lib/vcs/svn.js
var exec = require('child_process').exec,
	ArrayHelper = require('../util/arrayhelper'),
	arrayHelper = new ArrayHelper(),
	config = require('../../config'),
	Builder = require('../builder'),
	util = require('util');

function Svn() {

}

Svn.prototype.getDiffLog = function (fromBranch, toBranch, callback) {
	var builder = new Builder();
	var count = 0;

	this.getMinVersion(fromBranch, toBranch, function (minRevision) {
		this.getLog(fromBranch, minRevision, function (log) {
			builder.setOldLog(log);
			if (count++) {
				callback(builder.getDiff());
			}
		});

		this.getLog(toBranch, minRevision, function (log) {
			builder.setNewLog(log);
			if (count++) {
				callback(builder.getDiff());
			}
		});
	}.bind(this));
};

Svn.prototype.getLog = function (branch, minRevision, callback) {
	this._executeSvnCommand(util.format("log -g -r %s:HEAD %s%s", minRevision, config.vcs.path, branch),  function (err, stdout, stderr) {
		if (err) throw err;

		callback(stdout);
	});
};

Svn.prototype.getMinVersion = function (oldBranch, newBranch, callback) {
	this.getRevisionsForBranch(oldBranch, function (oldRevisions) {
		this.getRevisionsForBranch(newBranch, function (newRevisions) {
			var revisionDiff = arrayHelper.diff(newRevisions, oldRevisions);

			callback(Math.min.apply(null, revisionDiff) - 1);
		});
	}.bind(this));
};

Svn.prototype.getRevisionsForBranch = function (branch, callback) {
	this._executeSvnCommand("propget svn:mergeinfo " + config.vcs.path + branch, function (err, stdout, stderr) {
		if (err) throw err;

		var revisions = [];
		var branchInfos = stdout.split("\n");
		for (var i in branchInfos) {
			if (0 === branchInfos[i].length) continue;
			var branchData = branchInfos[i].split(":");
			var revisionData = branchData[1].split(",");
			for (var j in revisionData) {
				var revisionString = revisionData[j];
				if (revisionString.indexOf('-') === -1) {
					revisions.push(parseInt(revisionString));
				} else {
					var lowerEndRevisions = revisionString.split("-");
					for (var k = parseInt(lowerEndRevisions[0]); k <= lowerEndRevisions[1]; k++) {
						revisions.push(k);
					}
				}
			}

		}

		callback(revisions.sort(function (a, b) {
			return a - b;
		}));
	});
};


Svn.prototype._executeSvnCommand = function(command, callback) {
	var command_string = "svn " + command;

	if (config.vcs.user && config.vcs.password) {
		command_string += util.format(" --username=%s --password=%s", config.vcs.user, config.vcs.password);
	}

	exec(command_string, {maxBuffer: 1024 * 1024 * 100}, callback);
};

module.exports = Svn;
